### PR TITLE
11.0 imp sale credit point

### PR DESCRIPTION
**Pull request by simahawk**

### CODE REVIEW

File: sale_credit_point/models/point_history.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    _rec_name = "partner_id"
+
+    partner_id = fields.Many2one(
+        comodel_name="res.partner",
+        string="Partner",
+        required=True,
+    )
+    operation = fields.Selection(
+        string="Operation",
+        selection="_get_operation_selection",
```
Comment (simahawk) [inline sale_credit_point/models/point_history.py]: _selection_operation

File: sale_credit_point/models/partner.py [hunk 2]
```diff
@@ -14,6 +14,12 @@ class ResPartner(models.Model):
         readonly=True,
         default=0,
     )
+    yearly_point_increase = fields.Monetary(
+        string='Yearly points increase',
+        currency_field='credit_point_currency_id',
+        readonly=True,
+        compute="_compute_yearly_point_increase"
+    )
```
Comment (simahawk) [inline sale_credit_point/models/partner.py:27]: you miss an o2m to the history model here

File: sale_credit_point/models/partner.py [hunk 3]
```diff
@@ -63,3 +69,24 @@ def credit_point_increase(self, amount, comment=''):
...
+            'amount': amount,
+            'comment': comment,
+        }
+        return history_model.create(vals)
+
+    def _compute_yearly_point_increase(self):
+        history = self.env["credit.point.history"]
+        current_year = date.today().year
+        for record in self:
+            # TODO: is search_read better in such case?
```
Comment (simahawk) [inline sale_credit_point/models/partner.py]: yes, read only what you need. OR, as you are already bypassing ORM security, you can even use raw SQL query (you can even group by partner_id and do only one select too, if you prefer)

File: sale_credit_point/models/point_history.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        currency_field='credit_point_currency_id',
+        readonly=True,
+        default=0,
+        required=True,
+    )
+    credit_point_currency_id = fields.Many2one(
+        related="partner_id.credit_point_currency_id",
+        readonly=True,
+    )
+    execution_datetime = fields.Datetime(
```
Comment (simahawk) [inline sale_credit_point/models/point_history.py]: you don't need this field... you have: `create_date` ;)
If you want to display it in view you just have to re-declare it here to expose it. `create_date = fields.Datetime()` is enough

**Review by mpanarin — APPROVED** [review]
Great changes, I approve

File: sale_credit_point/models/point_history.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        string="Execution time",
+        default=lambda self: self._default_execution_datetime(),
+        required=True,
+    )
+    year = fields.Char(
```
Comment (simahawk) [inline sale_credit_point/models/point_history.py]: I don't think you need this either. You can [select dates by year](https://stackoverflow.com/questions/35704412/postgresql-query-current-year).
If you decide to keep it, make it an index.

**Review by simahawk — COMMENTED** [review]
also: where are tests? :)

**Comment by mpanarin** [comment]
@simahawk please update review )

File: sale_credit_point/models/point_history.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    credit_point_currency_id = fields.Many2one(
+        related="partner_id.credit_point_currency_id",
+        readonly=True,
+    )
+    create_date = fields.Datetime()
+    comment = fields.Char(
+        string="Comment"
+    )
+
+    def _selection_operation(self):
```
Comment (simahawk) [inline sale_credit_point/models/point_history.py]: move this to a `POINT_OPERATIONS` var into partner.py then import it in the wizard and here

File: sale_credit_point/tests/test_credit_history.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+from odoo import fields
+
+
+
+class TestHistory(TransactionCase):
+
+    def setUp(self):
+        super().setUp()
+        self.history_model = self.env["credit.point.history"]
+        self.partner = self.env['res.partner'].create({
```
Comment (simahawk) [inline sale_credit_point/tests/test_credit_history.py]: `.with_context(tracking_disable=True)` I know, I missed it in my tests, add it there too pls :wink:

File: sale_credit_point/tests/test_credit_history.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                'credit_point': 5,
+                'comment': 'test',
+            },
+        ]
+        for vals in wiz_vals:
+            wiz = self._get_wiz(**vals)
+            wiz.action_update_credit()
+
+        history = self.history_model.sudo().search([])
+        self.assertEqual(len(history), len(wiz_vals))
```
Comment (simahawk) [inline sale_credit_point/tests/test_credit_history.py]: this is another test :wink:

File: sale_credit_point/views/point_history.xml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                <field name="amount" readonly="True"/>
+                <field name="credit_point_currency_id" readonly="True"/>
+                <field name="credit_point_currency_id" readonly="True" invisible="1" />
+                <field name="create_date" readonly="True"/>
+                <field name="comment" readonly="True"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="credit_point_history_action" model="ir.actions.act_window">
```
Comment (simahawk) [inline sale_credit_point/views/point_history.xml]: what about filtering this by default on current year only and group  them by partner?

**Review by simahawk — COMMENTED** [review]
LGTM but:
1. I'd add filtering on history otherwise you'll have a bloated view
2. update the README w/ this new feature
3. add you to contribs ;)

File: sale_credit_point/models/point_history.py [hunk 10]
```diff
@@ -34,3 +35,14 @@ class PointHistory(models.Model):
     comment = fields.Char(
         string="Comment"
     )
+    current_year = fields.Boolean(
+        compute="_compute_current_year",
+        store=True,
+    )
+
+    @api.depends('create_date')
```
Comment (simahawk) [inline sale_credit_point/models/point_history.py]: so this is going to work only on create.... :)
You don't need this field. 1st example I grepped in odoo code:
```xml
<filter string="This Year" name="year" domain="[('create_date','&lt;=', time.strftime('%%Y-12-31')),('create_date','&gt;=',time.strftime('%%Y-01-01'))]"/>
```

**Review by simahawk — COMMENTED** [review]
:stop_sign: :smile:

**Pull request merged**